### PR TITLE
[td] Consistent pytest cache

### PR DESCRIPTION
**Pull request by clee2000**
Move the pytest cache downloading into the build step and store it in additional ci files so that it stays consistent during sharding.  

Only build env is taken into account now instead of also test config since we might not have the test config during build time, making it less specific, but I also think this might be better since tests are likely to fail across the same test config (I also think it might be worth not even looking at build env but thats a different topic)

Each cache upload should only include information from the current run.  Do not merge current cache with downloaded cache during upload (shouldn't matter anyways since the downloaded cache won't exist at the time)

From what I cant tell of the s3 retention policy, pytest cache files will be deleted after 30 days (cc @ZainRizvi to confirm), so we never have to worry about space or pulling old versions.

### CODE REVIEW

File: .github/workflows/_linux-build.yml [hunk 1]
```diff
@@ -118,6 +118,13 @@ jobs:
           test-matrix: ${{ inputs.test-matrix }}
           job-name: ${{ steps.get-job-id.outputs.job-name }}
 
+      - name: Download pytest cache
+        uses: ./.github/actions/pytest-cache-download
+        continue-on-error: true
+        with:
+          cache_dir: .pytest_cache
+          job_identifier: ${{ github.workflow }}_${{ inputs.build-environment }}
```
Comment (ZainRizvi) [inline .github/workflows/_linux-build.yml:126]: Will the build-environment during build time be enough info to detect all the test-time buld-environments?  Do all the test jobs who's caches we want share the same prefix?

File: tools/test/heuristics/test_heuristics.py [hunk 2]
```diff
@@ -6,10 +6,10 @@
 from typing import Any, Dict, Set
 from unittest import mock
 
-REPO_ROOT = pathlib.Path(__file__).resolve().parent.parent.parent
+REPO_ROOT = pathlib.Path(__file__).resolve().parent.parent.parent.parent
 try:
     # using tools/ to optimize test run.
-    sys.path.append(str(REPO_ROOT))
+    sys.path.insert(0, str(REPO_ROOT))
```
Comment (ZainRizvi) [inline tools/test/heuristics/test_heuristics.py:12]: Out of curiosity, why was this needed? And should this become our new, universal pattern?

File: tools/test/heuristics/test_heuristics.py [hunk 3]
```diff
@@ -6,10 +6,10 @@
 from typing import Any, Dict, Set
 from unittest import mock
 
-REPO_ROOT = pathlib.Path(__file__).resolve().parent.parent.parent
+REPO_ROOT = pathlib.Path(__file__).resolve().parent.parent.parent.parent
```
Comment (ZainRizvi) [inline tools/test/heuristics/test_heuristics.py:9]: Why did these changes require going up another level?

**Review by ZainRizvi — APPROVED** [review]

File: tools/test/heuristics/test_heuristics.py (hunk 3 above)
Comment (clee2000) [reply tools/test/heuristics/test_heuristics.py:9]: the old repo_root pointed at pytorch/tools isntead of pytorch, I don't know how it worked before this

File: tools/test/heuristics/test_heuristics.py (hunk 2 above)
Comment (clee2000) [reply tools/test/heuristics/test_heuristics.py:12]: This was originally me trying to figure out why I couldn't run the file, but we do this in run_test.py to ensure that we pick up the tools folder in pytorch and not the tools folder in some other repo, which someone encountered as an error becore

File: .github/workflows/_linux-build.yml (hunk 1 above)
Comment (clee2000) [reply .github/workflows/_linux-build.yml:126]: I think yes, I changed all `job_identifier`s to be this format so they should all be saving to this prefix, and the download doesn't have any of sha, test config or shard, so it should pull everything from the build envrionment.
